feat(Academic Recruitment): created Academic Research Recruitment Tab with working form

CheeMinHao opened the pull request:
Features Created:
Dummy Your Projects Tab
Dummy All Projects Tab
New projects Tab (connected to Firebase)

Blaise1030 commented on src/common/Tab.vue:
```diff
@@ -0,0 +1,29 @@
+<template>
+    <div class="tab-content pad--3" v-show="title == selectedTitle">
```
use `title === selectedTitle`

Blaise1030 commented on src/common/TabsWrapper.vue:
```diff
@@ -0,0 +1,65 @@
+            <li 
+                v-for="title in tabTitles" 
+                :key="title"
+                :class="{ selected: title == selectedTitle}"
```
use title === selectedTitle

Blaise1030 commented:
Also, attaching an image or video would be helpful

Blaise1030 commented on src/modules/navigation/SideNav.vue:
```diff
@@ -46,9 +58,9 @@ export default {
-        store.state.user_data 
-        ? store.state.user_data.roles.includes('admin') 
+    const userIsAdmin = computed(() =>
+        store.state.user_data
```
try  `store.state.user_data?.roles?.includes('admin')` instead of  `store.state.user_data ?store.state.user_data.roles.includes('admin') : false`

Blaise1030 commented on src/modules/recruitment/NewProject.vue:
```diff
@@ -0,0 +1,124 @@
+</template>
+
+<script>
+// import {defineComponent, reactive, computed} from 'vue'
```
Please remove unwanted imports and comments, to reduce cluter and improve readability

CheeMinHao commented:
Image - Nitin mentioned that it is quite difficult to implement independently, but will look into extracting from user's image, even so, it will be seen in the Project Details Page (not implemented yet)

Video - is there any reason that you can suggest for implementing this? Because usually research projects don't have a set prototype unless the staff would want to describe their projects through video.

Blaise1030 commented on src/store/index.ts:
```diff
@@ -7,21 +7,22 @@ import router from '@/router';
         isSideNavCollapsed: true,             // bool to check if sidenav is showing
         isLoading: true,                      // bool to keep track whether user is being retreived from the DB
-        is_under_maintenance: true,           // bool to know whether website is under maintenance and display maintenance screen
+        is_under_maintenance: false,          // bool to know whether website is under maintenance and display maintenance screen
```
Just asking, do we need `is_under_maintenance` turn to false in staging ?

CheeMinHao replied:
no idts, was curious why it's set in staging as well

Blaise1030 replied:
Okay, i think we can leave it as false

Blaise1030 commented on src/store/index.ts:
```diff
@@ -286,6 +287,16 @@ export default createStore({
+                .set(project).then(() => {
+                    state.projects.push(project)
+                }).catch(function(error) {
+                    console.log("Error getting document: " + error)
```
Try not to directly print out error messages for security reasons. try logging `Error getting document`or remove console.log

Blaise1030 commented:
@CheeMinHao What I meant was the video or image of your Pull Request, to let us see how your work is

NitinMa2 commented:
@Blaise1030 the changes can be viewed in the staging site. Saves the need to attach screenshots

NitinMa2 reviewed (commented):
I have quickly brushed through the code and here is my feedback as of now:

1. I am confused why the name you chose to go with is "Project" and not something more accurate like "Research". Is this intentional?
2. I would recommend you use Composition API with TypeScript instead of Options API with JS.
3. Please ensure all static styles are stored as variables.
4. When creating a new research project, there is no need for setting ID as the timestamp. Instead, Firebase will automatically auto-assign an ID and you can store the timestamp as an attribute.

All the best!

CheeMinHao commented:
Connected All Projects Tab with Firebase

To answer Nitin's comments:

1. I am confused why the name you chose to go with is "Project" and not something more accurate like "Research". Is this intentional? **(Yes it's intentional, "Research wouldn't make sense as an independent word hence the name "Project", would continue to look for better words due to its formality)**
2. I would recommend you use Composition API with TypeScript instead of Options API with JS. **(Will be resolved soon but would want to get the sub-features working first, as I'm still learning TS)**
3. Please ensure all static styles are stored as variables. **(Resolved)**
4. When creating a new research project, there is no need for setting ID as the timestamp. Instead, Firebase will automatically auto-assign an ID and you can store the timestamp as an attribute. **(Resolved)**

NitinMa2 commented:
Hi Min Hao,

Can we call it something like ResearchProjects then? For other developers, this name would allow them to associate the title with the research as Project could even mean student projects, etc. as we expand our scope of features.

About TS, you need not worry that much right now. It takes barely 2 minutes to convert a file to TS since the only things to be added are types. Try to make use of composition API if possible.. In the case you're more comfortable with options API, then its fine. I can help you refactor later.

CheeMinHao commented:
Changed file names

CheeMinHao commented:
**_Backlog 13/9:_**
- createProject and deleteProject function implemented in Blaise's API code base for research recruitment
           - Both functions can be found in **src/modules/recruitment/recruitmentAPI.ts**
           - involved web pages are **src/modules/recruitment/NewProject.vue** and **src/modules/recruitment/YourProject.vue**
- cleaned up old 'project' instance related code on **store/index.ts**

**_Bugs to mention_**
- For my account, if i remove the student role and I reload the page, it will regenerate the student role back on my firebase document. Not sure if it applies to other users.
- This bug will cause me to sometimes get 'All Projects' tab or some other times get 'Create Project', since I have both staff and student role.
- will consult @Blaise1030 on this matter

CheeMinHao commented:
**Backlog 16/9**

- updateProject implemented in Blaise's API code base for research recruitment and can be found in **src/modules/recruitment/recruitmentAPI.ts**
- Cleaned up some redundant code and console.logs
- Modified UI to allow project editing under **src/common/List.vue**, will require someone to maybe take up the task on polishing the button positioning

CheeMinHao commented:
**Current Bug**
After creating a project and reloading the page, "Create New Project" tab will be replaced with "All Projects" as the user has both staff and student roles. If an attempt to delete the role off of the database, the student role will regenerate after reload. Only when a new section tab is clicked (e.g. Events) and then clicking back to "Research" it will turn back to normal

https://user-images.githubusercontent.com/65438002/133885659-74431547-90df-42ed-93a5-32afa370cf81.mp4

CheeMinHao commented:
**Backlog 24/9**

- Fixed merge conflict while merging develop branch to research branch

NitinMa2 commented:
I've performed a review of all the FRs to ensure that everything is covered. I love how the feature is coming together with the UI, logic, and everything. I also noticed how long abstracts are trimmed in the list view. So cool! Here is some feedback I have for now:

1. Can we add form validation? It is not crucial for now but something that is important to eventually add.
2. Staff are not able to view the All Projects tab. We should consider showing this tab to them.
3. Actions such as edit, delete, reject, approve, etc. should have popups to confirm that the user really wants to perform the action. For example, "This action will delete the project permanently. Are you sure that you want to proceed?"
4. As a student, when I showed interest in one research project (and got rejected), all other research projects by that supervisor were no longer visible whether or not I have indicated an interest in them. There might be some issue with the logic in `!state.userInvolvements.find(findProject(p))` from my limited exploration.
5. When a document is being uploaded there is no user feedback on the screen to indicate that the upload is being processed. In my case, I was urged to press the upload button twice. 
6. The interest button is visible for a project that the student already registered their interest for. This could be a bit misleading. We could look into a way to improve that UX?
8. Can we improve the responsiveness of the pages?

We can look into these improvements for the time being. With regards to the FRs:

1. A feature to perform preliminary checks for the documents before it is sent to the supervisor was discussed but I'm not too sure how much it has been integrated into the platform. We could also get feedback from the stakeholders on whether that feature must be included in the coming deployment.
2. Notifications for important events, such as a new application, are yet to be added.
3. EOI form link is not integrated yet too.

CheeMinHao commented:
**Backlog 4/10**
- Integrated EOI Form as a button under **src/common/List.vue**. Would require some testing on the logic to ensure that the button is correctly displayed ONLY if the student has his/her interests approved.
